ddl/cache.lua

```diff
@@ -0,0 +1,136 @@
+local fiber = require('fiber')
+
+local cache = nil
+
+local CACHE_LOCK_TIMEOUT  = 3
+local SPACE_NAME_IDX      = 1
+local SHARD_FUNC_NAME_IDX = 2
+local SHARD_FUNC_BODY_IDX = 3
+
+-- Function decorator that is used to prevent cache_build() from being
+-- called concurrently by different fibers.
+local function locked(f)
+    return function(...)
+        local ok = cache.lock:put(true, CACHE_LOCK_TIMEOUT)
+
+        if not ok then
+            error("cache lock timeout is exceeded")
+        end
+
+        local status, err = pcall(f, ...)
+        cache.lock:get()
+
+        if not status or err ~= nil then
+            return err
+        end
+    end
+end
+
+-- Build cache.
+--
+-- Cache structure format:
+-- cache = {
+--   spaces = {
+--     'space_name' = {
+--       raw = {},       -- raw sharding metadata, used for ddl.get()
+--       processed = {}, -- table with parsed dot notation (like {'foo', 'bar'})
+--                       -- or a function ready to call (or a string with an error)
+--     }
+--   },
+--   lock,               -- locking based on fiber.channel()
+--   schema_version,     -- current schema version
+-- }
+--
+-- function returns nothing
+local cache_build = locked(function()
```

Comment:
Initially we discussed locking, because I was not sure, whether setting of a trigger yields / considered as DDL. I had no full information here, sorry. Anyway, locking around current implementation of the `cache_build()` function (without setting a trigger) is useless: `:pairs()` does not yield for memtx, all other operations are not about the database: just Lua object changes. Unlikely `_ddl_sharding_func` will be a vinyl space. But we can wrap it into `box.atomic()` to work correctly with vinyl too (it is harmless).

What about locking around listing / setting triggers? I looked over the code and found that this action just touch some runtime information. It does not yield. So, it seems, that locking here is not necessary.

Sorry for directing you by the wrong way.

Reply:
fixed, locked() removed. box.atomic is used instead